Complete Mod Manager

CoolSlimbo opened the pull request:
TODO: Add a visual end to it and make it generate scripts.rson when launching game.

SuperSlimey5 commented:
This would work as soon as Daddy Amos fixes the file system

SuperSlimey5 commented:
Fixed merge errors

Mauler125 commented on r5dev/filesystem/filesystem.h:
```diff
@@ -9,7 +9,7 @@
 {
 public:
-	void AddSearchPath(const char* pPath, const char* pPathID, SearchPathAdd_t addType);
+	void AddSearchPathP(const char* pPath, const char* pathID, SearchPathAdd_t addType);
```
Is there a reason this has been renamed?

Mauler125 commented on r5dev/sdklauncher/basepanel.cpp:
```diff
@@ -550,30 +783,173 @@ void CUIBaseSurface::ParseMaps()
+
+			if (smRegexMatches.size() > 0)
 			{
-				if (!this->m_MapCombo->Items.Contains("mp_lobby"))
```
The 'this->m_MapCombo->Items.Contains' has been removed here, but this has to be added back as the user will otherwise end up with duplicate map names if they have server and client VPK's installed.

Mauler125 commented on r5dev/sdklauncher/basepanel.cpp:
```diff
@@ -550,30 +783,173 @@ void CUIBaseSurface::ParseMaps()
+			}
+		}
+		catch (json::parse_error& e) {
+			// Error message for failure
```
Add the exception message.

Mauler125 commented on r5dev/sdklauncher/modManager.h:
```diff
@@ -0,0 +1,355 @@
+		for (auto& compMod : compiledMods) {
+			for (auto& script : compMod.customScripts) {
+				switch (script.runOn) {
+				case ui:
```
Switch order to "server, client, ui" instead.

Mauler125 commented on r5dev/sdklauncher/basepanel.cpp:
```diff
@@ -641,11 +1017,154 @@ void CUIBaseSurface::VirtualItemToClipboard(const std::unique_ptr<MouseEventArgs
+	case 1:
+		std::string text = "";
+		text = pSurface->m_ModList[pEventArgs->ItemIndex].m_object.name + " By ";
+		for (int i = 0; i < pSurface->m_ModList[pEventArgs->ItemIndex].m_object.authors.size(); i++) {
```
Cast the results of 'm_object.authors.size()' to int or (preferably) use size_t for 'i' instead of int. 'std::vector::size' returns a 'size_t'.

Mauler125 reviewed (changes requested):
Hi Slimbo, thanks for your commits. I have reviewed your changes, and found some issues or inconsistencies that need to be solved before this could be merged. See the comments added to the lines that need to be changed.

SuperSlimey5 commented:
Mainwindow's deleted because they weren't meant to be there

SuperSlimey5 reviewed (commented):
The commit will come through once I fix up modManager.h

CoolSlimbo commented:
- Time to explain everything, from the mod manager to the installer.

# Mod Manager
> A fully-featured mod manager, with easy-to-make mods, the ability to disable particular mods, along with all mods, and a GUI front for it.

### Features
- Addition to bottom of `launcher.exe`, allowing the user to select individual mods, view information relating to them, and toggle its state.
- A system to interpret mod folders. Mod Structure is outlined further below.
- Added a new launcher parameter to add the required `SearchPaths` for the mods to be read by the engine.
> This parameter is `-modded`, and is automatically added if the `Use Mods` box is checked.

#### Notes
> AppId is currently useless, besides a check I do. The **Further Additions** section will provide information on what it's planned to be used for.

### Mod Structure
- Below is the structure of a folder, and the contents of each folder when creating a mod to be used with this new mod manager system.
- [Sqaure Bracketed] text means a folder, {Curly Bracketed} text means a file, and text surrounded in angled brackets(<>) means that it's a placeholder name, not to be used when making a mod.

```
[mods]
└ [<Mod Name>]
    ├ [scripts]
    | └ <See "Scripts Folder" section>
    └ {mod.json}
      └ <See "Mod.json" section>
```

#### "\<Mod Name>" Explained
- The Mod Name section is the name of your mod. It does not have to match the name in `mod.json`, though it is ideal to name them the same thing. Spaces are allowed in the name but are not recommended. 
> **I.e.** if your mod is named `Spooderman Grappler`, name the folder `Spooderman-Grappler`

#### Scripts Folder
- The `scripts` folder contains any files that are being used by your mod. They are not all the files for it, only the required ones.
> **I.e.** if your mod replaces `mp_ability_grapple.txt` (located at `platform/scripts/weapons`), you would place that in `scripts/weapons` of the mod.
> 
> **Note:** Custom scripts are to be placed here, but they required further configuration in `mod.json`, outlined below.

#### Mod.json File
- The `mod.json` file holds all the important information about the mod. It is required for the mod manager to recognise the folder as a mod.
- Below is a table outlining elements in the `mod.json` file, what they are, and whether they are required:

|Key|Description|Required|
|---|-----------|--------|
|ApiId|Your unique mod identifier, useless as of right now.|Yes|
|Name|Sets the name of your mod in the launcher.|Yes|
|Description|Sets the subtitle of the mod in the launcher.|Yes|
|enabled|Controls whether the mod is enabled or not.|Yes|
|Authors|An array of strings detailing who made the mod.|No|
|Contacts|An array of strings detailing who to contact about the mod.|No|
|Version|A string of what version the mod is. Used primarily for the mod developer themselves.|No|
|CustomScripts|An array of CustomScript objects which tell the R5SDK how to load your scripts.|Yes if you have new script files|
> This was yoinked from the Icepick wiki, with changes to appropriate.
> The `enabled` field is lower case, not upper case. — If the field is missing it defaults to true. But should still be required.

> Below is the custom scripts information. The callback doesn't work as of right now.

|Key|Description|Required|
|---|-----------|--------|
|Path|Path to the script file to load, relative to your mods path.|Yes|
|RunOn|An rson string of which contexts to run the script in. eg. `CLIENT`|Yes|
|ClientCallback|The name of the function to load if loading on the Client context. eg. `Toolgun_Client_Init`|If RunOn specifies CLIENT|
|ServerCallback|The name of the function to load if loading on the Server context. eg. `Toolgun_Server_Init`|If RunOn specifies SERVER|

### Notes About The Mod Manager
- I might rewrite the mod manager in the future, because the code could likely be improved.
- Whilst the AppId isn't used as of right now, I plan on implementing a mod browser in the launcher, letting the user download the mod directly from a server.


# Updater
> Completed the SDK Updater. Allowing the `SDK Update` button to become enabled if need be. Allowing the installation of the latest SDK version. This also features an install game button, which runs if the game is missing files.

### Features
- `Update SDK` button becomes active if the user isn't running the latest version. The version is compared with a .cfg(a disguised `ini` file) located in `platform\cfg`. 
- If the user is missing any files part of the game, the `Launch Game` button becomes an `Install Game` button, which will connect to the torrent version of the file, and download any missing files.
> This will download only the missing files, saving bandwidth and space.

### Notes About The Updater
- After the `Updater SDK` features finish, it disables itself and updates the `launcher.cfg` file to reflect the new version.

# Additional Changes
- Moved and resized the console list view so that the scroll bars are hidden. This gives it a nicer look and finish.
> If a user doesn't have a scroll wheel, they shouldn't be playing Apex.
- Add a `fs::exists` check to a few files, saving from crash cases.
> These are the `vpk` folder check, and the `clean.bat` file.


## Some Words
- In the future I might make a custom file format for the mods, letting me integrate mods directly into the game.
> Something similar to Minecraft is what I have in mind.
- It might be worth in the future to let the user download the game from the US/EU server through the use of a shift click or something like that.

CoolSlimbo reviewed (commented):
Closed Review

Mauler125 commented on r5dev/sdklauncher/basepanel.cpp:
```diff
@@ -512,7 +939,98 @@ void CUIBaseSurface::CleanSDK(Forms::Control* pSender)
+
+	try {
+		lt::session ses;
+		lt::add_torrent_params p = lt::parse_magnet_uri("magnet:?xt=urn:btih:KCQJQT6DV2V4XWCOKCRM4EJELRLHQKI5&dn=R5pc_r5launch_N1094_CL456479_2019_10_30_05_20_PM&tr=udp%3A%2F%2Fwambo.club%3A1337%2Fannounce");
```
Don't hardcode magnet links.

Mauler125 commented on r5dev/sdklauncher/basepanel.cpp:
```diff
@@ -641,11 +1275,162 @@ void CUIBaseSurface::VirtualItemToClipboard(const std::unique_ptr<MouseEventArgs
+	pSurface->m_ManagerViewerNameLabel->SetText(object.name.c_str());
+	pSurface->m_ManagerViewerDescText->SetText(object.description.c_str());
+	std::string authors = "Authors: ";
+	for (int i = 0; i < object.authors.size(); i++) {
```
Cast to size_t or use size_t. You did not fix the issues requested by the first review, stopping current review here and requesting fixing everything from first review first before proceeding to second.

Mauler125 commented on .gitignore:
```diff
@@ -350,3 +350,8 @@ MigrationBackup/
 # Ionide (cross platform F# VS Code tools) working folder
 .ionide/
 .vscode/settings.json
+<<<<<<< HEAD
```
Properly fix merge conflicts...

Mauler125 commented on r5dev/filesystem/basefilesystem.cpp:
```diff
@@ -116,7 +116,7 @@ FileHandle_t CBaseFileSystem::VReadFromVPK(CBaseFileSystem* pFileSystem, FileHan
 		svFilePath.erase(0, 4);
 	}
-
+	
```
...

Mauler125 commented on r5dev/filesystem/filesystem.h:
```diff
@@ -9,7 +9,7 @@
 {
 public:
-	void AddSearchPath(const char* pPath, const char* pPathID, SearchPathAdd_t addType);
+	void AddSearchPath(const char* pPath, const char* pathID, SearchPathAdd_t addType);
```
See first review.

Mauler125 commented on r5dev/launcher/IApplication.cpp:
```diff
@@ -73,8 +74,8 @@ bool CModAppSystemGroup::Create(CModAppSystemGroup* pModAppSystemGroup)
 	}
-
 	g_bAppSystemInit = true;
+
```
...

Mauler125 commented on r5dev/launcher/launcher.cpp:
```diff
@@ -111,7 +113,7 @@ string LoadConfigFile(const string& svConfig)
 	else
 	{
-		spdlog::error("{:s}: '{:s}' does not exist!\n", __FUNCTION__, svConfig);
+		spdlog::error("{}: '{}' does not exist!\n", __FUNCTION__, svConfig);
```
...

Mauler125 commented on r5dev/sdklauncher/basepanel.cpp:
```diff
@@ -502,6 +701,234 @@ void CUIBaseSurface::Setup()
+// Purpose: Gets the requested config object. Each config object must be added to
+// v_LauncherValues so there's a default return value
+//-----------------------------------------------------------------------------
+std::string CUIBaseSurface::GetConfigCFG(std::string first) {
```
You are not modifying the string, pass by const reference instead.

Mauler125 commented on r5dev/sdklauncher/basepanel.cpp:
```diff
@@ -502,6 +701,234 @@ void CUIBaseSurface::Setup()
+// Purpose: Sets the requested config object - It also wasn't working just passing
+// cfg into json parser. So I gotta do this.
+//-----------------------------------------------------------------------------
+void CUIBaseSurface::SetConfigCFG(std::string first, std::string second) {
```
You are not modifying the 'first' string, pass by const reference instead.

Mauler125 commented on r5dev/sdklauncher/basepanel.cpp:
```diff
@@ -502,6 +701,234 @@ void CUIBaseSurface::Setup()
+		easy.perform();
+	}
+	catch (cURLpp::RuntimeError& e) {
+		// TODO: Error
```
Add the message.

Mauler125 commented on r5dev/sdklauncher/basepanel.cpp:
```diff
@@ -502,6 +701,234 @@ void CUIBaseSurface::Setup()
+		//pSurface->logText("Download Location: " + pSurface->sdkDownloadLocation);
+	}
+	catch (json::parse_error& e) {
+		// TODO: Error message
```
Add the message.

Mauler125 commented on r5dev/launcher/launcher.cpp:
```diff
@@ -7,6 +7,7 @@
 #include "core/stdafx.h"
 #include "tier0/commandline.h"
 #include "launcher/launcher.h"
+#include "../filesystem/filesystem.h"
```
Unnecessary include? See first review.

Mauler125 reviewed (changes requested):
There are still unresolved issues pending changes as requested by the first review. Reopening..

CoolSlimbo commented:
I’m prolly not gonna work on this for a while. If you do wanna build urself. You’ll need to do OpenSSL, and libtorrent urself due to the way they’re setup.
